cEP-0028.md

```diff
@@ -0,0 +1,428 @@
+
+This cEP describes the optimizations that need to be carried out
+to improve the performance of the coala based language server.
+
+## Introduction
+
+Language servers are the tools used to serve code editors with
+functionalities such as linting, code fixing, analysis etc in a
+standalone and independent manner. coala as a code style checker,
+linting tool and code fixer could be a powerful tool to be
+integrated into the code editor.
+
+Code Editing is a continuous process and the editors are expected
+to provide feedback on the fly, coala on the other hand was primarily
+designed as an analysis tool to be used after an incremental update to
+the codebase. Thus, the current implementation of a coala language
+server is severely limited by coala's performance.
+
+The performance of the server can be improved by primarily optimizing
+the coala and language server interface and improving the bear
+configuration options to avoid long running bears.
+
+## Proposal
+
+The optimization process focuses to maintain minimal changes to coala
+core and rather optimize the coala and language server interface. The
+following are the proposed ideas.
+
+1. Section Tagging<br/>
+   Some sections are not often required to be executed on every cycle of
+   coala analysis especially during use cases such as this. The parameter
+   set can thus be extended to support a parameter named `tags`. Since
+   these parameters are not associated with any particular bear but rather
+   control the effect of various sections during executing coala, no support
+   extension for bears would be necessary. These parameters enable support
+   for hooking a section for a particular tag type. For example:
+
+   ```
+   [longrunning]
+   tags = save
+   enabled = True
+   bears = SomeLongRunningBear
+   ```
+
+   This kind of a configuration will enable this section during analysis
+   cycles of coala via a language server when a save request is raised
+   while not impacting any untagged regular invocations.
+
+   ```
+   [longrunning]
+   files = *.py
+   bears = SomeLongRunningBear
+
+   [longrunning.some]
+   tags += save, change
+   bears = SomeLongRunningBear
+   ```
+
+   This example shows a case where the sections inherit the tags and
+   thereby `longrunning.some` runs on both `save` and `change` tagged
+   execution cycles.
+
+   To enable tagged executional support, coala-core needs to be extended
+   and at least support a new optional INPUT cli argument taking list of
+   tags to be executed. The execution of the sections will be based upon
+   the sections being tagged as at least one of the tags being invoked.
+   This extension of support for `tags` would require at least the
+   following changes:
+
+   - `coalib.settings.Section`<br/>
+     Adds a new method to the Section class to check if
+     the section should be enabled for a particular tag
+     based invocation.
+
+   ```python
+   class Section:
+
+     def is_enabled_for_tags(self, enabled_tags):
```

Comment:
This doesnt feel right to me.

I guess we need a `coalib.parsing.filters.TagFilter`

And maybe `Section` should expose the set of tags in an attribute.

Reply:
TagFilter sounds a better idea compared to this. Let's re-use the code. It saves dev & testing time

Reply:
And it will be easier to declare the tags for all the bears as well.

Reply:
"is_enabled" suggests this is doing a check against the `enabled` setting.  confusing..

maybe "has_all_tags(tag_list)" ?